### ISSUE TITLE
undefined reference to `rpl_fprintf'

### ISSUE THREAD

**Issue by Mingli-Yu** [akimd/bison#31]
Encounter below error when build bison, but actually I didn't find it calls rpl_fprintf in the function location_print, but it call fprintf instead, why does it complain "undefined reference to `rpl_fprintf'"? Thanks!

```
src/bison-location.o: In function `location_print':
location.c:(.text+0xae5): undefined reference to `rpl_fprintf'
location.c:(.text+0xb22): undefined reference to `rpl_fprintf'
location.c:(.text+0xb3d): undefined reference to `rpl_fprintf'
location.c:(.text+0xb65): undefined reference to `rpl_fprintf'
location.c:(.text+0xb80): undefined reference to `rpl_fprintf'
src/bison-location.o:location.c:(.text+0xbc4): more undefined references to `rpl_fprintf' follow
src/bison-getargs.o: In function `getargs':
getargs.c:(.text+0x1581): undefined reference to `rpl_printf'
getargs.c:(.text+0x18b1): undefined reference to `rpl_printf'
getargs.c:(.text+0x1928): undefined reference to `rpl_fprintf'
getargs.c:(.text+0x1d73): undefined reference to `rpl_printf'
getargs.c:(.text+0x1f63): undefined reference to `rpl_printf'
getargs.c:(.text+0x1f7f): undefined reference to `rpl_printf'
getargs.c:(.text+0x2037): undefined reference to `rpl_fprintf'
```

**Comment by akimd** [comment]
Hi.
I need the full logs of how you built Bison.  This function is defined in the `lib/` directory when it is needed.
Did you properly follow the procedure to bootstrap the package?

**Comment by Mingli-Yu** [comment]
gcc    -isystem/build/tmp/work/x86_64-linux/bison/3.5.3-r0/recipe-sysroot-native/usr/include -O2 -pipe  -L/build/tmp/work/x86_64-linux/bison/3.5.3-r0/recipe-sysroot-native/usr/lib                         -L/build/tmp/work/x86_64-linux/bison/3.5.3-r0/recipe-sysroot-native/lib                         -Wl,--enable-new-dtags                         -Wl,-rpath-link,/build/tmp/work/x86_64-linux/bison/3.5.3-r0/recipe-sysroot-native/usr/lib                         -Wl,-rpath-link,/build/tmp/work/x86_64-linux/bison/3.5.3-r0/recipe-sysroot-native/lib                         -Wl,-rpath,/build/tmp/work/x86_64-linux/bison/3.5.3-r0/recipe-sysroot-native/usr/lib                         -Wl,-rpath,/build/tmp/work/x86_64-linux/bison/3.5.3-r0/recipe-sysroot-native/lib                         -Wl,-O1 -Wl,--allow-shlib-undefined -Wl,--dynamic-linker=/buildarea3/cli10/wr-build/LINCD-1488/build/tmp-glibc/sysroots-uninative/x86_64-linux/lib/ld-linux-x86-64.so.2 -o src/bison src/bison-AnnotationList.o src/bison-InadequacyList.o src/bison-Sbitset.o src/bison-assoc.o src/bison-closure.o src/bison-complain.o src/bison-conflicts.o src/bison-derives.o src/bison-files.o src/bison-fixits.o src/bison-getargs.o src/bison-gram.o src/bison-graphviz.o src/bison-ielr.o src/bison-lalr.o src/bison-location.o src/bison-lr0.o src/bison-main.o src/bison-muscle-tab.o src/bison-named-ref.o src/bison-nullable.o src/bison-output.o src/bison-parse-gram.o src/bison-print-graph.o src/bison-print-xml.o src/bison-print.o src/bison-reader.o src/bison-reduce.o src/bison-relation.o src/bison-scan-code-c.o src/bison-scan-gram-c.o src/bison-scan-skel-c.o src/bison-state.o src/bison-symlist.o src/bison-symtab.o src/bison-tables.o src/bison-uniqstr.o           lib/libbison.a                                                                                                                                         src/bison-AnnotationList.o: In function `AnnotationList__debug':
AnnotationList.c:(.text+0x658): undefined reference to `rpl_fprintf'
AnnotationList.c:(.text+0x704): undefined reference to `rpl_fprintf'
AnnotationList.c:(.text+0x736): undefined reference to `rpl_fprintf'
AnnotationList.c:(.text+0x786): undefined reference to `rpl_fprintf'
AnnotationList.c:(.text+0x7ad): undefined reference to `rpl_fprintf'
src/bison-AnnotationList.o:AnnotationList.c:(.text+0x7fa): more undefined references to `rpl_fprintf' follow
collect2: error: ld returned 1 exit status
Makefile:3904: recipe for target 'src/bison' failed
make: *** [src/bison] Error 1

**Comment by akimd** [comment]
Hi.  No, I don't need just the call to the compiler, but the whole process.  How did you build bison?  Did you start from a tarball, or from the repository?

**Comment by Mingli-Yu** [comment]
Hi akimd,
Thanks very much for your respond!
Please help to check https://bugzilla.yoctoproject.org/show_bug.cgi?id=13825 for more details, thanks!

**Comment by akimd** [comment]
@Mingli-Yu Thanks for the pointer!  But I still lack information I need to understand what is going on.
1. What is the source for Bison?  Is this a released tarball?  Is this coming from a git repo?
2. How was Bison configured?  I do mean how was `./configure` run?
3. Is this a clean build?  Or for instance after a `git pull`?

Cheers!

**Comment by Mingli-Yu** [comment]
Hi akimd, 
Please check the info as below:
1, the source comes from https://ftp.gnu.org/gnu/bison/bison-3.5.3.tar.xz
2, ./configure  --build=x86_64-linux              --host=x86_64-linux             --target=x86_64-linux                   --prefix=/build/tmp/work/x86_64-linux/bison-native/3.5.3-r0/recipe-sysroot-native/usr                --exec_prefix=/build/tmp/work/x86_64-linux/bison-native/3.5.3-r0/recipe-sysroot-native/usr                   --bindir=/build/tmp/work/x86_64-linux/bison-native/3.5.3-r0/recipe-sysroot-native/usr/bin            --sbindir=/build/tmp/work/x86_64-linux/bison-native/3.5.3-r0/recipe-sysroot-native/usr/sbin                  --libexecdir=/build/tmp/work/x86_64-linux/bison-native/3.5.3-r0/recipe-sysroot-native/usr/libexec            --datadir=/build/tmp/work/x86_64-linux/bison-native/3.5.3-r0/recipe-sysroot-native/usr/share                 --sysconfdir=/build/tmp/work/x86_64-linux/bison-native/3.5.3-r0/recipe-sysroot-native/etc            --sharedstatedir=/build/tmp/work/x86_64-linux/bison-native/3.5.3-r0/recipe-sysroot-native/com                --localstatedir=/build/tmp/work/x86_64-linux/bison-native/3.5.3-r0/recipe-sysroot-native/var                 --libdir=/build/tmp/work/x86_64-linux/bison-native/3.5.3-r0/recipe-sysroot-native/usr/lib            --includedir=/build/tmp/work/x86_64-linux/bison-native/3.5.3-r0/recipe-sysroot-native/usr/include            --oldincludedir=/build/tmp/work/x86_64-linux/bison-native/3.5.3-r0/recipe-sysroot-native/usr/include                 --infodir=/build/tmp/work/x86_64-linux/bison-native/3.5.3-r0/recipe-sysroot-native/usr/share/info            --mandir=/build/tmp/work/x86_64-linux/bison-native/3.5.3-r0/recipe-sysroot-native/usr/share/man              --disable-silent-rules                  --disable-dependency-tracking                     M4=m4  --disable-nls
3, yes, it is a clean build.

**Comment by Mingli-Yu** [comment]
BTW,  I notice there is logic as "gl_cv_func_fprintf_posix=no" in config.log, why does it still try to find the gnulib replacement rpl_fprintf as the below error:
gcc    -isystem/build/tmp/work/x86_64-linux/bison/3.5.3-r0/recipe-sysroot-native/usr/include -O2 -pipe  -L/build/tmp/work/x86_64-linux/bison/3.5.3-r0/recipe-sysroot-native/usr/lib                         -L/build/tmp/work/x86_64-linux/bison/3.5.3-r0/recipe-sysroot-native/lib                         -Wl,--enable-new-dtags                         -Wl,-rpath-link,/build/tmp/work/x86_64-linux/bison/3.5.3-r0/recipe-sysroot-native/usr/lib                         -Wl,-rpath-link,/build/tmp/work/x86_64-linux/bison/3.5.3-r0/recipe-sysroot-native/lib                         -Wl,-rpath,/build/tmp/work/x86_64-linux/bison/3.5.3-r0/recipe-sysroot-native/usr/lib                         -Wl,-rpath,/build/tmp/work/x86_64-linux/bison/3.5.3-r0/recipe-sysroot-native/lib                         -Wl,-O1 -Wl,--allow-shlib-undefined -Wl,--dynamic-linker=/buildarea3/cli10/wr-build/LINCD-1488/build/tmp-glibc/sysroots-uninative/x86_64-linux/lib/ld-linux-x86-64.so.2 -o src/bison src/bison-AnnotationList.o src/bison-InadequacyList.o src/bison-Sbitset.o src/bison-assoc.o src/bison-closure.o src/bison-complain.o src/bison-conflicts.o src/bison-derives.o src/bison-files.o src/bison-fixits.o src/bison-getargs.o src/bison-gram.o src/bison-graphviz.o src/bison-ielr.o src/bison-lalr.o src/bison-location.o src/bison-lr0.o src/bison-main.o src/bison-muscle-tab.o src/bison-named-ref.o src/bison-nullable.o src/bison-output.o src/bison-parse-gram.o src/bison-print-graph.o src/bison-print-xml.o src/bison-print.o src/bison-reader.o src/bison-reduce.o src/bison-relation.o src/bison-scan-code-c.o src/bison-scan-gram-c.o src/bison-scan-skel-c.o src/bison-state.o src/bison-symlist.o src/bison-symtab.o src/bison-tables.o src/bison-uniqstr.o           lib/libbison.a                                                                                                                                         src/bison-AnnotationList.o: In function `AnnotationList__debug':
AnnotationList.c:(.text+0x658): undefined reference to `rpl_fprintf'
AnnotationList.c:(.text+0x704): undefined reference to `rpl_fprintf'
AnnotationList.c:(.text+0x736): undefined reference to `rpl_fprintf'
AnnotationList.c:(.text+0x786): undefined reference to `rpl_fprintf'
AnnotationList.c:(.text+0x7ad): undefined reference to `rpl_fprintf'
src/bison-AnnotationList.o:AnnotationList.c:(.text+0x7fa): more undefined references to `rpl_fprintf' follow
collect2: error: ld returned 1 exit status
Makefile:3904: recipe for target 'src/bison' failed
make: *** [src/bison] Error 1

**Comment by akimd** [comment]
Hi!
The problem might be simply that AnnotationList.c is missing `#include <stdio.h>`.  Could you please check if it addresses your problem?  Actually, let's address this directly in `src/system.h`:

``` diff
diff --git a/src/system.h b/src/system.h
index 606d3c46..86ddfe94 100644
--- a/src/system.h
+++ b/src/system.h
@@ -37,6 +37,7 @@
 
 # include <limits.h>
 # include <stddef.h>
+# include <stdio.h>
 # include <stdlib.h>
 # include <string.h>
 
```
Cheers!

**Comment by akimd** [comment]
I very much doubt my previous suggestion would change anything.  What do you have on this?
```
$ nm lib/libbison_a-fprintf.o
                 U ___error
                 U ___stack_chk_fail
                 U ___stack_chk_guard
                 U _free
                 U _fseterr
                 U _fwrite
0000000000000000 T _rpl_fprintf
                 U _vasnprintf
```

**Comment by Mingli-Yu** [comment]
$ nm libbison_a-fprintf.o
                 U __errno_location
0000000000000000 T fprintf
                 U free
                 U fseterr
                 U fwrite
                 U __stack_chk_fail
                 U vasnprintf

**Comment by akimd** [comment]
And is this also in `libbison.a`?
```
$ nm lib/libbison.a 2>/dev/null | grep fprintf
                 U _rpl_fprintf
                 U _rpl_fprintf
                 U _rpl_fprintf
                 U _rpl_fprintf
                 U _rpl_fprintf
                 U _rpl_fprintf
                 U _rpl_fprintf
                 U _rpl_fprintf
                 U _rpl_vfprintf
lib/libbison.a(libbison_a-fprintf.o):
0000000000000000 T _rpl_fprintf
                 U _rpl_fprintf
                 U _rpl_fprintf
                 U _fprintf
                 U _rpl_vfprintf
lib/libbison.a(libbison_a-vfprintf.o):
0000000000000000 T _rpl_vfprintf
```

**Comment by akimd** [comment]
(I mean, in a case of failure.  What I have shown is what is expected, but on my machine it works.  I'm interested in the failure)

**Comment by Mingli-Yu** [comment]
$ nm lib/libbison.a 2>/dev/null | grep fprintf
 U rpl_fprintf
                 U rpl_fprintf
                 U rpl_fprintf
                 U rpl_fprintf
                 U rpl_fprintf
                 U rpl_fprintf
                 U rpl_fprintf
libbison_a-fprintf.o:
0000000000000000 T fprintf
                 U rpl_fprintf
                 U rpl_vfprintf
libbison_a-vfprintf.o:
0000000000000000 T rpl_vfprintf

**Comment by akimd** [comment]
Well, I don't get it.  The symbol is present and the linker behaves as if it did not.  You'll have to track this down, it does not make any sense.

Are you sure you are _exactly_ in the same failure?

**Comment by akimd** [comment]
No wait!  You have `fprintf`, not `rpl_fprintf`!
Ok, let me try to make sense of this.

**Comment by akimd** [comment]
So maybe `lib/fprintf.o` was compiled too soon: `lib/stdio.h` was not generated yet, and therefore it did not have the `#define fprintf rpl_fprintf`, which is why the symbol is missing.  But `lib/stdio.h` _was_ generated when `AnnotationList.o` was compiled, so `AnnotationList.o` is looking for `rpl_fprintf`.

Could you:
1. report the content of the section about `fprintf` in `lib/stdio.h`.  E.g., I have this:
``` c
#if 1 || 1
# if (1 && 1) \
     || (1 && 0 && (0 || 0))
#  if !(defined __cplusplus && defined GNULIB_NAMESPACE)
#   define fprintf rpl_fprintf
#  endif
#  define GNULIB_overrides_fprintf 1
#  if 1 || 1
_GL_FUNCDECL_RPL (fprintf, int,
                  (FILE *restrict fp, const char *restrict format, ...)
                  _GL_ATTRIBUTE_FORMAT_PRINTF (2, 3)
                  _GL_ARG_NONNULL ((1, 2)));
#  else
_GL_FUNCDECL_RPL (fprintf, int,
                  (FILE *restrict fp, const char *restrict format, ...)
                  _GL_ATTRIBUTE_FORMAT_PRINTF_SYSTEM (2, 3)
                  _GL_ARG_NONNULL ((1, 2)));
#  endif
_GL_CXXALIAS_RPL (fprintf, int,
                  (FILE *restrict fp, const char *restrict format, ...));
# else
_GL_CXXALIAS_SYS (fprintf, int,
                  (FILE *restrict fp, const char *restrict format, ...));
# endif
# if __GLIBC__ >= 2
_GL_CXXALIASWARN (fprintf);
# endif
#endif
#if !1 && defined GNULIB_POSIXCHECK
# if !GNULIB_overrides_fprintf
#  undef fprintf
# endif
/* Assume fprintf is always declared.  */
_GL_WARN_ON_USE (fprintf, "fprintf is not always POSIX compliant - "
                 "use gnulib module fprintf-posix for portable "
                 "POSIX compliance");
#endif
```

2. Check that you can fix a failure by `touch lib/fprintf.c && make`.

Thanks in advance!

**Comment by Mingli-Yu** [comment]
$ ls -lt --full-time ./build/lib/libbison_a-fprintf.o
-rw-r--r-- 1 clt users 2208 2020-03-25 11:59:19.297890870 +0800 ./build/lib/libbison_a-fprintf.o
$ ls -lt --full-time ./build/lib/stdio.h
-rw-r--r-- 1 clt users 72770 2020-03-25 11:59:19.269889888 +0800 ./build/lib/stdio.h
$ ls -lt --full-time ./build/src/bison-AnnotationList.o
-rw-r--r-- 1 clt users 17024 2020-03-25 11:59:19.645903068 +0800 ./build/src/bison-AnnotationList.o

# the content of the section about fprintf in lib/stdio.h in my env.
#if 1 || 1
# if (1 && 1) \
     || (1 && 0 && (0 || 0))
#  if !(defined __cplusplus && defined GNULIB_NAMESPACE)
#   define fprintf rpl_fprintf
#  endif
#  define GNULIB_overrides_fprintf 1
#  if 1 || 1
_GL_FUNCDECL_RPL (fprintf, int,
                  (FILE *restrict fp, const char *restrict format, ...)
                  _GL_ATTRIBUTE_FORMAT_PRINTF (2, 3)
                  _GL_ARG_NONNULL ((1, 2)));
#  else
_GL_FUNCDECL_RPL (fprintf, int,
                  (FILE *restrict fp, const char *restrict format, ...)
                  _GL_ATTRIBUTE_FORMAT_PRINTF_SYSTEM (2, 3)
                  _GL_ARG_NONNULL ((1, 2)));
#  endif
_GL_CXXALIAS_RPL (fprintf, int,
                  (FILE *restrict fp, const char *restrict format, ...));
# else
_GL_CXXALIAS_SYS (fprintf, int,
                  (FILE *restrict fp, const char *restrict format, ...));
# endif
# if __GLIBC__ >= 2
_GL_CXXALIASWARN (fprintf);
# endif
#endif
#if !1 && defined GNULIB_POSIXCHECK
# if !GNULIB_overrides_fprintf
#  undef fprintf
# endif
/* Assume fprintf is always declared.  */
_GL_WARN_ON_USE (fprintf, "fprintf is not always POSIX compliant - "
                 "use gnulib module fprintf-posix for portable "
                 "POSIX compliance");
#endif

**Comment by akimd** [comment]
So AFAICT:
- `lib/stdio.h`'s timestamp shows its generation was ended before that of `libbison_a-fprintf.o`, would give the impression that all is well
- yet your `stdio.h` clearly shows that it `#define fprintf rpl_fprintf`
- so `libbison_a-fprintf.o` should be about `rpl_fprintf`, but it's about `fprintf` according to your `nm`.
So please, could you run the test I suggested:

> 2. Check that you can fix a failure by touch lib/fprintf.c && make.

Cheers!

**Comment by Mingli-Yu** [comment]
Thanks akimd!
As the failure doesn't occure every time but sometimes, I will try to remove libbison_a-fprintf.o and re-make it if it occurs next time.

**Comment by Mingli-Yu** [comment]
Hi akimd,
I want to send a patch to improve the parallel build issue related to "undefined reference to `rpl_fprintf'",
but I didn't find the correspond code ./lib/gnulib.mk in the git repo, but instead it exist in the source of
https://ftp.gnu.org/gnu/bison/bison-3.6.tar.xz. So could you help to guide how to send the patch, based on  the git repo or https://ftp.gnu.org/gnu/bison/bison-3.6.tar.xz. If the latter, where can I send the patch?
Thanks,

**Comment by akimd** [comment]
Hi!
This file is actually generated from... a generated file.  Please, just send the diff you suggest for the generated file, I'll see how to adjust its generation.

Thanks!

**Comment by Mingli-Yu** [comment]
I ever sent the patch to bison-patches@gnu.org, but I didn't find at https://lists.gnu.org/archive/html/bison-patches/2020-05/threads.html. Anyway, please check the diff as below(the diff is based on https://ftp.gnu.org/gnu/bison/bison-3.6.tar.xz):

diff --git a/lib/gnulib.mk b/lib/gnulib.mk
index 5ce2a5c..4ac4f78 100644
--- a/lib/gnulib.mk
+++ b/lib/gnulib.mk
@@ -121,6 +121,8 @@ lib_libbison_a_SOURCES =
 lib_libbison_a_LIBADD = $(gl_LIBOBJS)
 lib_libbison_a_DEPENDENCIES = $(gl_LIBOBJS)
 EXTRA_lib_libbison_a_SOURCES =
+$(lib_libbison_a_OBJECTS): $(BUILT_SOURCES) $(MOSTLYCLEANFILES:%.h)
+$(gl_LIBOBJS): $(BUILT_SOURCES) $(MOSTLYCLEANFILES:%.h)
 
 lib_libbison_a_CPPFLAGS = $(AM_CPPFLAGS) -DDEFAULT_TEXT_DOMAIN=\"bison-gnulib\"
 
diff --git a/src/local.mk b/src/local.mk
index c072d99..89321dd 100644
--- a/src/local.mk
+++ b/src/local.mk
@@ -24,6 +24,7 @@ if RELOCATABLE_VIA_LD
   src_bison_LDFLAGS = `$(RELOCATABLE_LDFLAGS) $(bindir)`
 endif
 
+$(src_bison_OBJECTS): $(BUILT_SOURCES) $(MOSTLYCLEANFILES:%.h)
 src_bison_CFLAGS = $(AM_CFLAGS) $(WERROR_CFLAGS)
 src_bison_SOURCES =                             \
   src/AnnotationList.c

**Comment by Mingli-Yu** [comment]
The patch also at https://lists.gnu.org/archive/html/bison-patches/2020-05/msg00055.html

**Comment by akimd** [comment]
I believe I have finally nailed it down.  See https://lists.gnu.org/r/bug-bison/2020-09/msg00003.html.  I just released 3.7.2 with the fix.

Thanks for your patience!
